### ISSUE TITLE
Improvements to zio.Chunk

### ISSUE THREAD

**Issue by vasilmkd** [zio/zio#2225]
# TL; DR;

I would like to start a discussion on how to introduce specialized `Chunk` types for primitive types that are going to be exported, so that "power" users can utilize them in performance sensitive code.

I have documented my explorations, anyone who is interested to read is most welcome to. I will tag as many people as I can think of that might be interested in this subject area. I apologize if I have left someone out and feel free to tag them.

@jdegoes @iravid @regiskuckaertz @adamgfraser @mschuwalow @mijicd @ioleo 

# Long version

This issue was first inspired by #1479, or implementing Chunks that are backed by `java.nio.Buffer`s, which would allow processing of buffer data without copying it into an array first.

The curious thing about `java.nio.Buffer` and more importantly, its subtypes (which are the ones that we are interested in, `ByteBuffer`, `IntBuffer` etc...) is that, for performance reasons (avoiding boxing of primitive values), the base `java.nio.Buffer` abstract class does not provide methods for reading from the buffer. Instead, those methods are available only on the concrete buffer types. You can consult the `Buffer` implementation [here](https://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html).

Because of the uncertain future of primitive specialization in Scala 3, some time ago, the `@specialized` annotation was completely removed from the `ZIO` codebase, and it took any autogenerated performance optimizations when it comes to handling of primitive types with it.

Now, before we go any further, it is important to keep in mind that the high-level, functional API of `zio.Chunk`, along with its prevalent use of combinators that accept higher-order functions will negate any too clever special implementations for specific primitive types. A good example is `Byte`, a primitive type that already sees a lot of usages in Streams, and would be extremely useful in libraries like `zio-nio` and probably to the users of `zio-kafka` and future users of `zio-http`.

However, while higher order functions impose performance overhead for primitive types, it doesn't mean that the other combinators that do not use higher order functions should not benefit from performance improvements. Case in point, the internal implementation of `Chunk` is heavily based on the index based lookup method `def apply(n: Int): A`, which is a polymorphic method, and therefore always boxes. To somewhat mitigate this issue, the `Arr` implementation of `Chunk` exists which overrides most of the combinators to directly fetch and write data to the backing array of the `Chunk`, without going through the polymorphic `apply` method. This is great and already works with arrays of primitive values.

In a discussion with @iravid, we came to the conclusion that we would like to offer a way for interested users to pattern match on a `Chunk`, see that it is a `ByteChunk` (which would be a special implementation of `Chunk[Byte]` that would expose a `def getByte(n: Int): Byte` method, i.e. a non-polymorphic method for extracting primitive values which avoids boxing.

And everything would have been fine, we would have implemented a `ByteArr`, `IntArr` and other subtypes of `Arr`, that also expose a `getByte`, `getInt` methods and go on with our days.

Here again comes the need to have `java.nio.Buffer` backed chunks. And for these chunks too, we would have liked to offer this `getByte` method for `ByteBuffer` backed chunks, `getInt` method for `IntBuffer` backed chunks, etc. Apart from the obvious desire to abstract over these "byte oriented data structure" backed chunks (with data structure being an array of bytes, buffer of bytes, even a singleton byte), all of them offering the `getByte(n: Int): Byte` method, there is, what I personally consider to be a code smell, the issue of publicly exposing tens of classes so that the user can pattern match on them `(Byte, Int, Long, Short, Float, Double, Boolean) x (Arr, Buffer, Singleton)` (the cross product of the sets of primitive types and chunk backing types, with more chunk types coming in the future).

So, in order to avoid this explosion of exported classes, I was thinking that this situation requires a rethinking of the class hierarchy of Chunk.

There are two approaches that I could think of.

The first one is, to retrofit the new methods to the current class hierarchy, by letting some classes extend new traits that we will introduce for each of the primitive types. Let me illustrate that with an example:

```scala
trait ByteChunk {
  def getByte(n: Int): Byte
}

case class ByteArr(array: Array[Byte]) extends Arr[Byte] with ByteChunk {}

case class ByteBuff(buffer: ByteBuffer) extends Buff[Byte] with ByteChunk {}
```

where `sealed trait Buff[A]` can contain overrides for common Chunk combinators, to improve performance, in a way that makes sense for `java.nio.Buffer`s, similar to `Arr`. This implementation should have the performance of an interface call. (Although, trying this out in a toy example compiles down to `invokevirtual` calls, not `invokeinterface`, but I am far from an expert on this.)

Of course, we should be able to add extractor methods to all of the companion objects of the new traits, so that the users can pattern match nicely, instead of doing type checks.

The second approach is to bifurcate Chunks into a structure class hierarchy and behavior class hierarchy. Let me try to illustrate what I mean by that. The structure class hierarchy would consist of the `ByteChunk`, `IntChunk`, other primitive types and `RefChunk[A]` for reference types. These, could then mix in behavior, depending on their backing (`Arr`, `Buff`, `Singleton`, which would contain optimized overrides for the specific behavior of the given chunk, and these might be implemented by mixin traits). This approach is probably more controversial, because of the added complexity that mixins introduce, but also because of the megamorphic dispatch that most likely happens in this case. Again, I am no expert on JIT compilers. Unfortunately, avoiding mixins in this case leads to a lot of code duplication, which whoever is interested, can take a look at [here](https://github.com/vasilmkd/zio/blob/ee511b768de20df6830aea818a32e6b359cc7acf/core/shared/src/main/scala/zio/Chunk2.scala).

A third approach, but probably not very popular for `ZIO` would be typeclasses. This is just an idea, I haven't explored this. At all.

Here, I'm leaving this issue slightly open ended, in hope that a discussion will come out of it. I will totally understand if everything that is written here is complete nonsense and should be abandoned, I won't feel like it has been wasted effort at all. I can only say that I would really like to see `java.nio.Buffer` based `Chunks` become a reality without too much code duplication, and hopefully with reasonable performance.

**Comment by iravid** [comment]
Great write-up @vasilmkd! I'm not a fan of the complexity of the second approach. The first approach is pretty promising, I think, and even though it uses mixins, it'd be interesting to benchmark the performance hit compared to the current implementation.

**Comment by vasilmkd** [comment]
Technically, a Scala mixin is a trait with no abstract methods, that is used in the with clause of the class definition in order to add or override specific behavior of the class. The second approach outlined above does use mixins, and I agree, I'm not a huge fan either.

The first approach is not a mixin technically, because the `ByteChunk` trait is a simple interface and provides no default implementation. I'm interested to hear from others, what the performance implications of this are.

I'm only writing this to properly define what I refer to in my issue description.

**Comment by iravid** [comment]
Ah, I see what you mean. The implementations of all the combinators would need to be re-implemented in `Buff`?

**Comment by vasilmkd** [comment]
But materializing the `Buff` is the same as now, it requires copying the contents of the `Buffer`. WDYT?

**Comment by iravid** [comment]
Yeah, but that's the best you can do with a buffer as it is mutable and cannot be used to back an immutable structure like Chunk.

**Comment by vasilmkd** [comment]
True, but I think we are misunderstanding each other. self.materialize would create an intermediate array which is then flatmapped.

**Comment by iravid** [comment]
Yes, I understand. The tradeoff here is between saving that copy and re-implementing all the combinators. I would benchmark to see if the performance gain is worthwhile as a function of the chunk size.

**Comment by vasilmkd** [comment]
What if, we override only some combinators that are clear wins, and inherit the others from the base `Chunk` trait? I don't envision `Buff` to be as complex as `Arr` to be honest.

**Comment by iravid** [comment]
If you copy the data out of a buffer into an array, that's basically the performance of `System.arraycopy`. It is going to be very fast, and the performance of every combinator on chunk will be dominated not by that copy but by the performance of boxing/unboxing incurred by applying the user's function to the data.

So again, let's benchmark. I would wager the difference is immaterial, particularly at larger (and more realistic) chunk sizes.

**Comment by vasilmkd** [comment]
I see. Well if the copying is happening at native speeds, than maybe we don't need this discussion at all. Why not, for now, add convenience constructors for `java.nio.Buffer`s and pick this up again post 1.0?

**Comment by hearnadam** [comment]
I suspect this won't be pursued due to binary compatibility. Feel free to reopen if you think otherwise.
